Test suite now uses socket pairs instead of memory streams

martinschroeder opened the pull request:
~~This PR contains 2 changes:~~

~~Warnings when using undefined signal constants in a PHPUnit data provider are avoided. Missing constants for signals are defined as needed in the abstract test case.~~

Streams used by test cases are created as connected socket pairs. Most event loop extension do not support arbitrary file descriptors but they all support socket pairs. The created socket pairs use unix domain sockets if they are available. The fallback to INET sockets is needed in order to support Windows.

martinschroeder commented:
There is one failing test (testIgnoreRemovedCallback) remaining with lib ev loop. I think that this test is problematic because it assumes that the loop backend triggers events in the same order as writes within the test code.

martinschroeder commented:
The latest commit fixed the ordering problem with libev. The build failed due to a timing issue with the HHVM build, can't do anything about this...

@WyriHaximus this PR should fix some tests failing due to invalid file descriptors in [reactphp-async-interop-loop
](https://github.com/WyriHaximus/reactphp-async-interop-loop) as well.

clue commented:
> The build failed due to a timing issue with the HHVM build, can't do anything about this...

I've restarted the HHVM build and this seems to have fixed this for now :+1:

martinschroeder commented:
Splitting it up sounds good to me.
Should i modify this PR to just address the socket pair stuff and create a new PR for the signal handling later? You could squash to commits during the merge to get rid of the signal stuff.

martinschroeder commented:
I removed the signal-related part of this PR. Using real sockets in tests enables IO testing using the abstract test class in #25 and #12.

clue commented on tests/AbstractLoopTest.php:
```diff
@@ -16,11 +16,39 @@ public function setUp()
+    /**
+     * @deprecated Use createSocketPair() instead.
+     */
     public function createStream()
```
Should we remove this altogether?

martinschroeder replied:
Removing the method would be a better solution. I just marked it as deprecated to keep it backward compatible.

clue replied:
Can these be removed now? Afaict they're not needed anymore and BC doesn't really apply to our test suite because it's not part of our public API :+1: (see also below)

clue commented on tests/AbstractLoopTest.php:
```diff
@@ -16,11 +16,39 @@ public function setUp()
+        if ($this->isUnixSocketSupported()) {
+            $sockets = stream_socket_pair(STREAM_PF_UNIX, STREAM_SOCK_STREAM, STREAM_IPPROTO_IP);
+        } else {
+            $sockets = stream_socket_pair(STREAM_PF_INET, STREAM_SOCK_STREAM, STREAM_IPPROTO_IP);
```
How can I test this? Shouldn't this use `STREAM_IPPROTO_TCP`? Do we need both branches?

martinschroeder replied:
I could turn this into one branch like this:
```php
$domain = $this->isUnixSocketSupported() ? STREAM_PF_UNIX : STREAM_PF_INET;
$sockets = stream_socket_pair($domain, STREAM_SOCK_STREAM, STREAM_IPPROTO_IP);
```
This ensures that it is allways the same function call.

clue commented on tests/AbstractLoopTest.php:
```diff
@@ -16,11 +16,39 @@ public function setUp()
+        }
+        
+        foreach ($sockets as $socket) {
+            if (function_exists('stream_set_read_buffer')) {
```
This check could be performed before the loop. However, this is lacking some documentation, why is this needed in the first place? And why do we not bother if it doesn't exist?

martinschroeder replied:
Disabling the read buffers is necessary because some event loop extensions will not report streams as readable if PHP does read buffering. Setting the buffer to 0 will ensure that all backends will work on the socket directly.

clue commented on tests/AbstractLoopTest.php:
```diff
@@ -16,11 +16,39 @@ public function setUp()
+    /**
+     * @deprecated Use createSocketPair() and fwrite() instead.
+     */
     public function writeToStream($stream, $content)
```
Should we remove this altogether?

clue commented on tests/AbstractLoopTest.php:
```diff
@@ -167,71 +195,76 @@ public function emptyRunShouldSimplyReturn()
+        $this->loop->nextTick(function () {
+            $this->loop->stop();
+        });
+        
```
Unfortunately this PR contains plenty of unrelated formatting changes (see above and below) which make this kind of hard to review. Can you revert these changes so we can focus on what's important here? Thanks!

martinschroeder replied:
I will have to start from scratch to undo the formatting. I will look into it later today or maybe tomorrow.

martinschroeder commented:
I fixed the code formatting. Unfortunately the HHVM build failed again due to a timing problem...

jsor commented:
@martinschroeder You should probably rebase/merge the master branch to get b946301bfc0f3996855c5f1f8298c04d2701349c in which should fix the HHVM tests.

martinschroeder commented:
I merged the master branch, this fixed the HHVM test case. 👍

clue commented:
Ping @martinschroeder, what do you think about the above suggestion?

Also, may I ask you to rebase this on the new "0.4" branch so we get this in for the upcoming v0.4.3 release? :shipit:

martinschroeder commented:
I rebased the PR on the 0.4 branch and removed the obsolete / deprecated methods `createStream()` and `writeToStream()`. I squashed all my commits into a single commit to avoid work-in-progress commits in the history.

clue commented:
Awesome, much appreciated! :+1:

WyriHaximus commented:
Looking good, thanks 👍
